alacritty_terminal/src/term/cell.rs

```diff
@@ -47,7 +50,244 @@ impl<T: Copy> ResetDiscriminant<T> for T {
 
 impl ResetDiscriminant<Color> for Cell {
     fn discriminant(&self) -> Color {
-        self.bg
+        self.colors.bg()
+    }
+}
+
+bitflags! {
+    pub struct ColorTy : u8 {
+        const NAMED   = 0b0000_0000;
+        const INDEXED = 0b0000_0001;
+        const SPEC    = 0b0000_0010;
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+struct ColorMask {
+    mask: u8,
+}
+
+impl ColorMask {
+    /// The offset in bits for `bg` color type in mask.
+    const BG_OFFSET: u8 = 2;
+    /// Whether the underline was set.
+    const HAS_UNDERLINE: u8 = 0b1000_0000;
+    /// Reset type mask.
+    const RESET_MASK: u8 = 0b0000_0011;
+    /// The offset in bits for `fg` color type in mask.
+    const UNDERLINE_OFFSET: u8 = 4;
+
+    #[inline]
+    fn new() -> Self {
+        Default::default()
+    }
+
+    #[inline]
+    fn fg_ty(&self) -> ColorTy {
+        ColorTy::from_bits(self.mask & Self::RESET_MASK).unwrap()
+    }
+
+    #[inline]
+    fn bg_ty(&self) -> ColorTy {
+        ColorTy::from_bits(self.mask >> Self::BG_OFFSET & Self::RESET_MASK).unwrap()
+    }
+
+    #[inline]
+    fn underline_ty(&self) -> ColorTy {
+        ColorTy::from_bits(self.mask >> Self::UNDERLINE_OFFSET & Self::RESET_MASK).unwrap()
+    }
+
+    #[inline]
+    fn set_fg_ty(&mut self, ty: ColorTy) {
+        self.mask &= !(Self::RESET_MASK);
+        self.mask |= ty.bits();
+    }
+
+    #[inline]
+    fn set_bg_ty(&mut self, ty: ColorTy) {
+        self.mask &= !(Self::RESET_MASK << Self::BG_OFFSET);
+        self.mask |= ty.bits() << Self::BG_OFFSET;
+    }
+
+    #[inline]
+    fn set_underline_ty(&mut self, ty: ColorTy) {
+        self.mask |= Self::HAS_UNDERLINE;
+        self.mask &= !(Self::RESET_MASK << Self::UNDERLINE_OFFSET);
+        self.mask |= ty.bits() << Self::UNDERLINE_OFFSET;
+    }
+
+    #[inline]
+    fn clear_underline(&mut self) {
+        self.mask &= !Self::HAS_UNDERLINE;
+    }
+
+    #[inline]
+    fn has_underline(&self) -> bool {
+        self.mask & Self::HAS_UNDERLINE != 0
+    }
+}
+
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
+pub union PackedColor {
```

Comment:
You mentioned on IRC you had an idea on how to clean this up, but I don't see any significant difference? This 250 lines of garbage is still something I can't really warm up with.